web/api/v1/langserver.go

```diff
@@ -0,0 +1,134 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus-community/promql-langserver/prometheus"
+	"github.com/prometheus-community/promql-langserver/rest"
+	"github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+func NewLangServer(externalURL string, q storage.Queryable,
+	tr func(context.Context) TargetRetriever, logger log.Logger) (http.Handler, error) {
+	// create the custom prometheus client dedicated to the langserver
+	langserverClient := newPrometheusLangServerClient(externalURL, newWebkit(q, tr))
+	// TODO use the method CreateInstHandler once it is taking the interface prometheus.Registerer in parameter
+	// instead of the struct Registry
+	return rest.CreateHandler(context.Background(), langserverClient, logger)
+}
+
+type prometheusLangServerClient struct {
+	prometheus.Client
+	externalURL string
+	webkit      *webkit
+}
+
+func newPrometheusLangServerClient(externalURL string, webkit *webkit) prometheus.Client {
+	return &prometheusLangServerClient{
+		externalURL: externalURL,
+		webkit:      webkit,
+	}
+}
+
+// Metadata returns the first occurrence of metadata about metrics currently scraped by the metric name.
+func (c *prometheusLangServerClient) Metadata(ctx context.Context, metric string) (v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, metric, 1)
+	if len(metadata) == 0 {
+		return v1.Metadata{}, nil
+	}
+	return v1.Metadata{
+		Type: v1.MetricType(metadata[metric][0].Type),
+		Help: metadata[metric][0].Help,
+		Unit: metadata[metric][0].Unit,
+	}, nil
+}
+
+// AllMetadata returns metadata about metrics currently scraped for all existing metrics.
+func (c *prometheusLangServerClient) AllMetadata(ctx context.Context) (map[string][]v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, "", -1)
+	result := make(map[string][]v1.Metadata)
+	for metric, infos := range metadata {
+		result[metric] = make([]v1.Metadata, 0, len(infos))
+		for i, info := range infos {
+			result[metric][i] = v1.Metadata{
+				Type: v1.MetricType(info.Type),
+				Help: info.Help,
+				Unit: info.Unit,
+			}
+		}
+	}
+	return result, nil
+}
+
+// LabelNames returns all the unique label names present in the block in sorted order.
+// If a metric is provided, then it will return all unique label names linked to the metric during a predefined period of time
+func (c *prometheusLangServerClient) LabelNames(ctx context.Context, metricName string) ([]string, error) {
+	if len(metricName) == 0 {
+		names, _, err := c.webkit.labelNames(ctx, minTime, maxTime)
+		return names, err
+	}
+	labelNames, _, err := c.webkit.series(ctx,
+		[][]*labels.Matcher{
+			{
+				&labels.Matcher{
+					Type:  labels.MatchEqual,
+					Name:  "__name__",
```

Comment:
Please avoid any additional uses of model, especially for accessing constants. Limit it to time parsing stuff.

Reply:
but that's quite clean to use a constant. I hardcoded the string here because I didn't know about this constant.
It limited the number of error if the value is changing. But if it's really what you want, then I will keep the harcoded value.

Reply:
@brian-brazil What's the problem with using `model`?

Reply:
It's largely a historical thing where various things were attempted to be over-formalised into separate types, code that should have been in one repo ended up split across two, and generally getting in the way. There's a few things where there is actually common code (e.g. duration parsing), but the rest should go away and live in the actual repo that cares about it.

Another example in this PR is using a special type for labelnames/values, where the current practice is to use a simple string.

Reply:
I don't feel strongly about the whole `model` repo or not, but still having this as a constant then somewhere in the Prometheus repo would make sense at least. But in the interest of not opening up that larger discussion (which would involve replacing the existing `model.MetricNameLabel` everywhere, I'm fine having it as a string literal here for now.

Reply:
the literal `__name__` will stay like that until there is a consensus regarding the usage of the package `model`